autoarray/plot/mat_wrap/mat_plot.py

```diff
@@ -0,0 +1,754 @@
+
+        - `Array`:, using `plt.imshow`.
+        - `Grid`: using `plt.scatter`.
+        - `Line`: using `plt.plot`, `plt.semilogy`, `plt.loglog` or `plt.scatter`.
+        - `VectorField`: using `plt.quiver`.
+        - `RectangularMapper`: using `plt.imshow`.
+        - `VoronoiMapper`: using `plt.fill`.
+
+        Parameters
+        ----------
+        units : mat_wrap.Units
+            The units of the figure used to plot the data structure which sets the y and x ticks and labels.
+        figure : mat_wrap.Figure
+            Opens the matplotlib figure before plotting via `plt.figure` and closes it once plotting is complete
+            via `plt.close`
+        cmap : mat_wrap.Cmap
+            Customizes the colormap of the plot and its normalization via matplotlib `colors` objects such 
+            as `colors.Normalize` and `colors.LogNorm`.
+        colorbar : mat_wrap.Colorbar
+            Plots the colorbar of the plot via `plt.colorbar` and customizes its appearance and ticks using methods
+            like `cb.set_yticklabels` and `cb.ax.tick_params`.
+        tickparams : mat_wrap.TickParams
+            Customizes the appearances of the y and x ticks on the plot, (e.g. the fontsize), using `plt.tick_params`.
+        yticks : mat_wrap.YTicks
+            Sets the yticks of the plot, including scaling them to new units depending on the `Units` object, via
+            `plt.yticks`.
+        xticks : mat_wrap.XTicks
+            Sets the xticks of the plot, including scaling them to new units depending on the `Units` object, via
+            `plt.xticks`.
+        title : mat_wrap.Title
+            Sets the figure title and customizes its appearance using `plt.title`.        
+        ylabel : mat_wrap.YLabel
+            Sets the figure ylabel and customizes its appearance using `plt.ylabel`.
+        xlabel : mat_wrap.XLabel
+            Sets the figure xlabel and customizes its appearance using `plt.xlabel`.
+        legend : mat_wrap.Legend
+            Sets whether the plot inclues a legend and customizes its appearance and labels using `plt.legend`.
+        output : mat_wrap.Output
+            Sets if the figure is displayed on the user's screen or output to `.png` using `plt.show` and `plt.savefig`
+        """
+
+        super().__init__(
+            units=units,
+            figure=figure,
+            cmap=cmap,
+            colorbar=colorbar,
+            tickparams=tickparams,
+            yticks=yticks,
+            xticks=xticks,
+            title=title,
+            ylabel=ylabel,
+            xlabel=xlabel,
+            legend=legend,
+            output=output,
+        )
+
+        self.line_plot = line_plot
+
+        self.for_subplot = False
+
+    def plot_line(
+        self,
+        y,
+        x,
+        label=None,
+        plot_axis_type="semilogy",
+        vertical_lines=None,
+        vertical_line_labels=None,
+    ):
+
+        if y is None:

```

Comment:
Should raise an exception instead? Presumably if someone calls this function they wouldn't expect it to just do nothing

Reply:
Wrapping matplotlib we basically may or may not call a series of matplotlib.pyplot methods. So these if None's are so that we can call these methods all in sequuence but they only get used if the features is on. Raising an exception would break this.

I don't think theres a better way to handle this? I guess I'm not expecitng users to directly call these methods. May I should preceed them with a double underscore?

Reply:
It would be better to decide whether to call the function based on a feature flag than the function deciding not to work internally based on the absence of some data